### DISCUSSION TITLE
Note on frozen components

### DISCUSSION THREAD

**Discussion by fcggamou** [explosion/spaCy#7305] [Help: Best practices]
Hi, this is a very minor question, but this block of text in the documentation sparked some doubts:

> "NOTE ON FROZEN COMPONENTS
> Even though frozen components are not updated during training, they will still run during training and evaluation. This is very important, because they may still impact your model’s performance – **for instance, a sentence boundary detector can impact what the parser or entity recognizer considers a valid parse**. So the evaluation results should always reflect what your pipeline will produce at runtime."

Can you clarify the bolded text? I'm using NER models; should I add a sentence boundary detector to my pipelines? If I do, how can I make the NER use it? 

Thanks!

**Comment by polm** [comment]
For reference, [here](https://spacy.io/usage/training#config-components:~:text=NOTE%20ON%20FROZEN%20COMPONENTS)'s a link to the note.

The note is not a recommendation that you use a particular component, it's just a note that even though frozen components don't learn themselves, they can modify the input (and thus the performance) of components that come after them in the pipeline. A sentence boundary detector can modify the parser or entity recognizer because parses won't cross sentence boundaries - an adjective can't modify a noun in another sentence. 

The takeaway from this is that your frozen components should be **consistent** between training and inference, not that you should use a particular component.

> I'm using NER models; should I add a sentence boundary detector to my pipelines? If I do, how can I make the NER use it?

By default spaCy models include sentence tokenization so you don't have to do anything. If you want to customize it look at the [Sentencizer](https://spacy.io/api/sentencizer) (rule-based) or [SentenceRecognizer](https://spacy.io/api/sentencerecognizer) (statistical, trainable) components.

**Comment by fcggamou** [comment]
Thanks for your answer. 

This is more of an advice kind of question, inspired by that note: when training a new NER model, is it recommended to use a sentence recognizer first? I.e. should my training examples be per sentence, instead as per text document?

Thanks!

**Reply by polm** [reply]
I would generally use a sentence tokenizer of some kind, yes. Generally models work better with shorter sequences - think sentences as opposed to whole documents - and typically an entity crossing sentence boundaries would be an error.

**Reply by polm** [reply]
>  I.e. should my training examples be per sentence, instead as per text document?

Ah, actually, to clarify - using a sentence recognizer doesn't mean that your training examples have to be isolated sentences. Your training data can include whole documents, but spaCy can split them into sentences and process them while retaining some document context. This is why the Annotation data format supports supplying sentence data relative to a document.

https://spacy.io/api/data-formats#dict-input

**Reply by fcggamou** [reply]
Thanks a lot for your input. 

> but spaCy can split them into sentences and process them while retaining some document context. This is why the Annotation data format supports supplying sentence data relative to a document.

So to clarify, if I annotate my data with sentence data e.g. processing my training examples with a sentencizer, the training process of my NER will automatically make use of this?

**Reply by polm** [reply]
Yes, if you provide sentence boundaries in your training data it can be used by a Scorer to calculate per-sentence metrics, for example.

**Reply by adrianeboyd** [reply]
In v3.0, the answer is sort of. The NER component itself won't see the sentence boundaries while updating its model, but the evaluation on the dev set shown by `spacy train` does include the sentence boundaries when the NER model is run for scoring purposes.

In v3.1 we're planning to add a setting so you can choose whether you want the NER model to see the sentence boundaries or not while updating the model.

We usually recommend paragraph-ish documents for training. Very large documents aren't beneficial (the NER model only looks at relatively local features around each token) and it's better to be able to shuffle the training examples more. Of course, it's always the case that it's best if your training data matches your task, so if you expect to only have single-sentence texts as input, then it would make sense to train on single-sentence examples.

**Reply by fcggamou** [reply]
Thanks a lot for the detailed answers. 

> In v3.1 we're planning to add a setting so you can choose whether you want the NER model to see the sentence boundaries or not while updating the model.

Sounds like a very nice feature!
